modules/text/src/text_detectorCNN.cpp

```diff
@@ -0,0 +1,458 @@
+#include "precomp.hpp"
+#include "opencv2/imgproc.hpp"
+#include "opencv2/core.hpp"
+
+
+
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <queue>
+#include <algorithm>
+#include <iosfwd>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+
+#ifdef HAVE_CAFFE
+#include "caffe/caffe.hpp"
+#endif
+
+#ifdef HAVE_DNN
+#include "opencv2/dnn.hpp"
+#endif
+
+using namespace cv::dnn;
+
+#define CV_WARN(message) fprintf(stderr, "warning: %s (%s:%d)\n", message, __FILE__, __LINE__)
+
+namespace cv { namespace text {
+
+inline bool fileExists (String filename) {
+    std::ifstream f(filename.c_str());
+    return f.good();
+}
+
+class DeepCNNTextDetectorCaffeImpl: public DeepCNNTextDetector{
```

Comment:
If `HAVE_CAFFE` macro is not defined isn't the whole `DeepCNNTextDetectorCaffeImpl` class not of use? Shouldn't you just envelope all of the class definition with the macro instead of just sections?

Reply:
In constructor there is an else section which throws an exception. I think this is better as constructor will be available anyway.